formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt

```diff
@@ -0,0 +1,300 @@
+            is CborFloat -> encoder.encodeSerializableValue(CborFloatSerializer, value)
+            is CborInt -> encoder.encodeSerializableValue(CborIntSerializer, value)
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): CborPrimitive<*> {
+        val result = decoder.asCborDecoder().decodeCborElement()
+        if (result !is CborPrimitive<*>) throw CborDecodingException("Unexpected CBOR element, expected CborPrimitive, had ${result::class}")
+        return result
+    }
+}
+
+/**
+ * Serializer object providing [SerializationStrategy] and [DeserializationStrategy] for [CborNull].
+ * It can only be used by with [Cbor] format an its input ([CborDecoder] and [CborEncoder]).
+ */
+internal object CborNullSerializer : KSerializer<CborNull>, CborSerializer {
+
+    override val descriptor: SerialDescriptor =
+        buildSerialDescriptor("kotlinx.serialization.cbor.CborNull", SerialKind.ENUM)
+
+    override fun serialize(encoder: Encoder, value: CborNull) {
+        val cborEncoder = encoder.asCborEncoder()
+        cborEncoder.encodeTags(value.tags)
+        encoder.encodeNull()
+    }
+
+    override fun deserialize(decoder: Decoder): CborNull {
+        decoder.asCborDecoder()
+        if (decoder.decodeNotNullMark()) {
+            throw CborDecodingException("Expected 'null' literal")
+        }
+
+        decoder.decodeNull()
+        return CborNull()
+    }
+}
+
+
+internal object CborIntSerializer : KSerializer<CborInt>, CborSerializer {
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor("kotlinx.serialization.cbor.CborInt", PrimitiveKind.LONG)
+
+    override fun serialize(encoder: Encoder, value: CborInt) {
+        val cborEncoder = encoder.asCborEncoder()
+        cborEncoder.encodeTags(value.tags)
+        when (value.sign) {
+            //@formatter:off
+            CborInt.Sign.ZERO, CborInt.Sign.POSITIVE -> cborEncoder.encodePositive(value.value)
+            CborInt.Sign.NEGATIVE                    -> cborEncoder.encodeNegative(value.value)
+            //@formatter:on
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): CborInt {
+        val result = decoder.asCborDecoder().decodeCborElement()
+        if (result !is CborInt) throw CborDecodingException("Unexpected CBOR element, expected CborInt, had ${result::class}")
+        return result
+    }
+}
+
+internal object CborFloatSerializer : KSerializer<CborFloat>, CborSerializer {
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor("kotlinx.serialization.cbor.CborDouble", PrimitiveKind.DOUBLE)
+
+    override fun serialize(encoder: Encoder, value: CborFloat) {
+        val cborEncoder = encoder.asCborEncoder()
+        cborEncoder.encodeTags(value.tags)
+        encoder.encodeDouble(value.value)
+    }
+
+    override fun deserialize(decoder: Decoder): CborFloat {
+        decoder.asCborDecoder()
+        return CborFloat(decoder.decodeDouble())
```

Comment:
Tags will be lost here during deserialization (the same is true for CborMap and CborList deserialization routines).